man/regconvert.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "REGCONVERT" "1" "10 June 2026" "" "The Regina Handbook"

.SH NAME
regconvert \- Convert between different Regina file formats
.SH SYNOPSIS

\fBregconvert\fR [ \fB-2\fR | \fB-3\fR ] [ \fB-x\fR | \fB-u\fR ] \fB\fIold-file\fB\fR [ \fB\fInew-file\fB\fR ]


\fBregconvert\fR { \fB-v, --version\fR | \fB-?, --help\fR }

.SH "DESCRIPTION"
.PP
This utility converts back and forth between different Regina
file formats.
.PP
Regina's file formats include:
.TP
\fBThird-generation XML\fR
This is the modern XML format introduced in Regina\~7.0.
Older versions of Regina (6.0.1 and below) cannot read it at all.
.TP
\fBSecond-generation XML\fR
This is the original XML format that was used in
Regina\~3.0-6.0.1 (from 2002-2021).
Current versions of Regina can still read and write this format
if required.
.TP
\fBFirst-generation binary format\fR
This is the ancient binary format that was used in Regina\~2.x
(from 2000-2002).
Support for this format was dropped completely in September\~2013,
and modern versions of Regina cannot read or write it.

If you still have ancient files that use this format, you will need to
find a copy of Regina\~3.0-4.93 (from 2002-2012),
which can convert these old binary files to the second-generation
XML format.
.PP
The argument \fIold-file\fR should be the file to
read and convert; the argument \fInew-file\fR should
be the name of the new converted file to write.  These two filenames
may be the same, in which case the old file will be replaced with the new.
.PP
If the argument \fInew-file\fR is missing then the
new file will be written to standard output, which forces the output to be
uncompressed XML (see option \fB-u\fR).
.SH "OPTIONS"
.TP
\fB-2\fR
Convert to the older second-generation XML format,
as used in Regina\~3.0-6.0.1.
.TP
\fB-3 (default)\fR
Convert to the modern third-generation XML format,
introduced in Regina\~7.0.
.TP
\fB-x (default)\fR
Write the output as compressed XML\&.
.TP
\fB-u\fR
Write the output as uncompressed (plain text) XML\&.
.TP
\fB-v, --version\fR
Show which version of Regina is being used, and exit
immediately.
.TP
\fB-?, --help\fR
Display brief usage information, and exit immediately.
.SH "MACOS USERS"
.PP
If you downloaded a drag-and-drop app bundle, this utility is
shipped inside it.  If you dragged Regina to the main
Applications folder, you can run it as
/Applications/Regina.app/Contents/MacOS/regconvert\&.
.SH "WINDOWS USERS"
.PP
The command-line utilities are installed beneath the
\fIProgram\~Files\fR directory; on some
machines this directory is called
\fIProgram\~Files\~(x86)\fR\&.
You can start this utility by running
c:\\Program\~Files\\Regina\\Regina\~8.0\\bin\\regconvert.exe\&.
.SH "SEE ALSO"
.PP
regfiletype,
regina-gui\&.
.SH "AUTHOR"
.PP
This utility was written by Benjamin Burton
<bab@maths.uq.edu.au>\&.
Many people have been involved in the development
of Regina; see the users' handbook for a full list of credits.


utils/regconvert.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Convert old-style binary data files to new-style XML                  *
 *                                                                        *
 *  Copyright (c) 1999-2026, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  As an exception, when this program is distributed through (i) the     *
 *  App Store by Apple Inc.; (ii) the Mac App Store by Apple Inc.; or     *
 *  (iii) Google Play by Google Inc., then that store may impose any      *
 *  digital rights management, device limits and/or redistribution        *
 *  restrictions that are required by its terms of service.               *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public License     *
 *  along with this program. If not, see <https://www.gnu.org/licenses/>. *
 *                                                                        *
 **************************************************************************/

#include "packet/packet.h"
#include "utilities/exception.h"
#include <cstdlib>
#include <cstring>

void usage(const char* progName, const std::string& error = std::string()) {
    if (! error.empty())
        std::cerr << error << "\n\n";

    std::cerr << "Usage:\n";
    std::cerr << "    " << progName << " [ -2 | -3 ] [ -x | -u ]"
        << " <old-file> [ <new-file> ]\n"
        "    " << progName << " [ -v, --version | -?, --help ]\n";
    std::cerr << std::endl;
    std::cerr << "    -2 : Convert to the old second-generation format used by Regina 3.0-6.0.1\n";
    std::cerr << "    -3 : Convert to the current third-generation file format (default)\n";
    std::cerr << std::endl;
    std::cerr << "    -x : Convert to compressed XML (default)\n";
    std::cerr << "    -u : Convert to uncompressed XML\n\n";
    std::cerr << "    -v, --version : Show which version of Regina "
        "is being used\n";
    std::cerr << "    -?, --help    : Display this help\n";
    std::cerr << std::endl;
    std::cerr << "    <new-file> may be the same as <old-file>.\n";
    std::cerr << "    <new-file> defaults to standard output (implies -u).\n";
    exit(1);
}

int main(int argc, char* argv[]) {
    std::string oldFile, newFile;
    char typeOpt = 0;
    char versionOpt = 0;

    // Parse command line.
    char optChar;
    for (int i = 1; i < argc; i++) {
        if (*argv[i] == '-') {
            // Standard arguments:
            if (strcmp(argv[i], "-?") == 0 || strcmp(argv[i], "--help") == 0)
                usage(argv[0]);
            if (strcmp(argv[i], "-v") == 0 ||
                    strcmp(argv[i], "--version") == 0) {
                if (argc != 2)
                    usage(argv[0], "Option --version cannot be "
                        "used with any other arguments.");
                std::cout << PACKAGE_BUILD_STRING << std::endl;
                exit(0);
            }

            // Option.
            if (! argv[i][1])
                usage(argv[0], std::string("Invalid option: ") + argv[i]);
            else if (argv[i][2])
                usage(argv[0], std::string("Invalid option: ") + argv[i]);

            // Argument has length 2.
            optChar = argv[i][1];
            if (optChar == 'x' || optChar == 'u') {
                if (typeOpt)
                    usage(argv[0],
                        "More than one file type has been specified.");
                else
                    typeOpt = optChar;
            } else if (optChar == '2' || optChar == '3') {
                if (versionOpt)
                    usage(argv[0],
                        "More than one file format version has been specified.");
                else
                    versionOpt = optChar;
            } else
                usage(argv[0], std::string("Invalid option: ") + argv[i]);
        } else if (*argv[i]) {
            // Filename.
            if (oldFile.empty())
                oldFile = argv[i];
            else if (newFile.empty())
                newFile = argv[i];
            else
                usage(argv[0], "More than two files have been specified.");
        } else
            usage(argv[0], "Empty arguments are not allowed.");
    }

    if (oldFile.empty())
        usage(argv[0], "No files have been specified.");

    // Add default options.
    if (! typeOpt)
        typeOpt = (newFile.empty() ? 'u' : 'x');
    regina::FileFormat version = (versionOpt == '2' ?
        regina::FileFormat::XmlGen2 : regina::FileFormat::XmlGen3);

    // Check we're allowed to use stdout if we've asked for it.
    if (newFile.empty() && typeOpt != 'u')
        usage(argv[0],
            "Only uncompressed XML can be written to standard output.");

    // Read the old file.
    std::shared_ptr<regina::Packet> tree;
    try {
        tree = regina::open(oldFile.c_str());
    } catch (const regina::FileError&) {
        std::cerr << "File " << oldFile << " could not be read.\n";
        return 1;
    } catch (const regina::InvalidInput&) {
        std::cerr << "File " << oldFile
            << " does not appear to be a Regina data file.\n";
        return 1;
    }

    // Write the new file.
    if (newFile.empty()) {
        // Standard output
        if (typeOpt == 'u') {
            tree->writeXMLFile(std::cout, version);
        } else {
            std::cerr << "File <stdout> could not be written.\n";
            return 1;
        }
    } else try {
        // Real output file.
        // Use compressed / uncompressed XML
        tree->save(newFile.c_str(), typeOpt == 'x', version);
    } catch (const regina::FileError&) {
        std::cerr << "File " << newFile << " could not be written.\n";
        return 1;
    }

    return 0;
}

